torchgeo/trainers/simclr.py

```diff
@@ -0,0 +1,259 @@
+            else:
+                state_dict = get_weight(weights).get_state_dict(progress=True)
+            self.backbone = utils.load_state_dict(self.backbone, state_dict)
+
+        # Create projection head
+        input_dim = self.backbone.num_features
+        if hidden_dim is None:
+            hidden_dim = input_dim
+        if output_dim is None:
+            output_dim = input_dim
+
+        self.projection_head = SimCLRProjectionHead(
+            input_dim, hidden_dim, output_dim, layers
+        )
+
+        # Define loss function
+        self.criterion = NTXentLoss(temperature, memory_bank_size, gather_distributed)
+
+        # TODO
+        # v1+: add global batch norm
+        # v2: add selective kernels, channel-wise attention mechanism
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Forward pass of the model.
+
+        Args:
+            x: Mini-batch of images.
+
+        Returns:
+            Output from the model.
+        """
+        h = self.backbone(x)
+        z = self.projection_head(h)
+        return cast(Tensor, z)
+
+    def training_step(self, batch: dict[str, Tensor], batch_idx: int) -> Tensor:
+        """Compute the training loss and additional metrics.
+
+        Args:
+            batch: The output of your DataLoader.
+            batch_idx: Integer displaying index of this batch.
+
+        Returns:
+            The loss tensor.
+        """
+        x = batch["image"]
+
+        in_channels = self.hparams["in_channels"]
+        assert x.size(1) == in_channels or x.size(1) == 2 * in_channels
+
+        if x.size(1) == in_channels:
+            x1 = x
+            x2 = x
+        else:
+            x1 = x[:, :in_channels]
+            x2 = x[:, in_channels:]
+
+        with torch.no_grad():
+            x1 = self.augmentations(x1)
+            x2 = self.augmentations(x2)
+
+        z1 = self(x1)
+        z2 = self(x2)
+
+        loss = self.criterion(z1, z2)
+
+        self.log("train_loss", loss)
+
+        return cast(Tensor, loss)
+
+    def validation_step(self, batch: dict[str, Tensor], batch_idx: int) -> None:
+        """No-op, does nothing."""
+
+    def test_step(self, batch: dict[str, Tensor], batch_idx: int) -> None:
+        """No-op, does nothing."""
+        # TODO
+        # v2: add distillation step
+
+    def predict_step(self, batch: dict[str, Tensor], batch_idx: int) -> None:
+        """No-op, does nothing."""
+
+    def configure_optimizers(self) -> tuple[list[Optimizer], list[LRScheduler]]:
+        """Initialize the optimizer and learning rate scheduler.
+
+        Returns:
+            Optimizer and learning rate scheduler.
+        """
+        # Original paper uses LARS optimizer, but this is not defined in PyTorch
+        optimizer = Adam(
+            self.parameters(),
+            lr=self.hparams["lr"],
+            weight_decay=self.hparams["weight_decay"],
+        )
+        lr_scheduler = CosineAnnealingLR(
```

Comment:
Why is `T_max` the total number of epochs? (I've seen this elsewhere, just don't know why it is)

Reply:
See the mathematical definition [here](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingLR.html). $T_{cur}$ is the number of epochs since the start (restarts are not implemented). Once $T_{cur}$ reaches $T_{max}$, the right half of the equation cancels out and the learning rate $\eta_t$ becomes $\eta_{min}$, which is 0.

Reply:
I know what it is, I don't know why it defaults to the total number of epochs (i.e. one cycle through the cosine)

Reply:
It's only a half-cycle ($[0, \pi]$). Convergence isn't guaranteed unless learning rate decays to zero. If you did a full-cycle, learning rate would go back to the initial learning rate. You could cycle multiple times and still end up at 0, but I think that's the point of warm restarts.